File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    
int answer = 0;
    map<string, int> ans;
    int q = 0;
    for (string str : want) {
        ans[str] = number[q++];
    }

    for (int i = 0; i <= discount.size() - 10; i++) {
        map<string, int> ans_copy = ans;
        for (int j = 0; j < 10; j++) {
            if (ans_copy.find(discount[i + j]) != ans_copy.end()) {
                ans_copy[discount[i + j]]--;
            }
        }

        bool isMatch = true;
        for (auto& pair : ans_copy) {
            if (pair.second > 0) {
                isMatch = false;
                break;
            }
        }

        if (isMatch) {
            answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    answer = 0
    
    # want 리스트의 제품들을 dictionary로 변환하여 원하는 수량 저장
    want_dict = dict(zip(want, number))
    
    # discount 리스트를 10일씩 슬라이딩하면서 확인
    for i in range(len(discount) - 9):
        # 현재 10일간의 할인 품목들을 카운트
        current_dict = {}
        for j in range(10):
            item = discount[i + j]
            current_dict[item] = current_dict.get(item, 0) + 1
        
        # 원하는 제품의 수량과 일치하는지 확인
        is_possible = True
        for item, count in want_dict.items():
            if current_dict.get(item, 0) != count:
                is_possible = False
                break
        
        if is_possible:
            answer += 1
            
    return answer